mesonbuild/wrap/wrap.py

```diff
@@ -566,18 +573,24 @@ def get_file_internal(self, what: str) -> str:
 
             return path.as_posix()
 
+    def unpack_archive(self, path):
+        with tempfile.TemporaryDirectory() as tmpdir:
+            shutil.unpack_archive(path, tmpdir)
+            src = glob.glob(f'{tmpdir}{os.path.sep}*{os.path.sep}')
+            if len(src) != 1:
+                raise WrapPatchFormatException("Error patch archive format.")
```

Comment:
If I understand correctly, this assumes the tarballs contains a single toplevel directory and we need to copy what it contains, right? `os.listdir(tmpdir)` seems more straightforward.

But that's not the format we were assuming before, won't it break the case where the tarball does not have a toplevel directory and must be extracted directly onto the subproject?

Reply:
> But that's not the format we were assuming before

Probably, no one actually uses or used this behaviour: current `~master` unpacks patch files from these archives without directories into `subprojects/` directly.

(Manual about patch archives extracting also can be understood in two ways, as for me)

Reply:
oh we used to extract in `subprojects/` and not `subprojects/foo`, ok. Wondering why we were catching exceptions and then extract into a tempdir, do you now what kind of exceptions we could have there?

Reply:
I don't know Python :-)

Reply:
You can raise WrapException() here, ideally with a more specific error message, saying we expected the archive to only contain a single directory.

Reply:
The exception causes it to unpack into a tempdir, and then invoke self.copy_tree which is an "enhanced copy" that also tries to chmod existing read-only files before copying a new version of the file. shutil.unpack_archive would just error out by default.

This has nothing to do with changing the extraction directory name.